marlowe-dashboard-client/src/Contact/View.purs

```diff
@@ -0,0 +1,113 @@
+module Contact.View
+  ( renderContacts
+  , renderNewContact
+  , renderContact
+  ) where
+
+import Prelude hiding (div)
+import Contact.Lenses (_key, _nickname)
+import Contact.Types (Action(..), Contact)
+import Contact.Validation (keyError, nicknameError)
+import Css (buttonClasses, classNames, h2Classes, textInputClasses, toggleWhen)
+import Data.Array (head, sort)
+import Data.Lens (view)
+import Data.Maybe (Maybe(..))
+import Halogen.HTML (HTML, button, div, h2, hr_, input, li, p_, span, text, ul)
+import Halogen.HTML.Events (onClick, onValueInput)
+import Halogen.HTML.Properties (InputType(..), disabled, placeholder, type_, value)
+import Material.Icons as Icon
+
+renderContacts :: forall p. Array Contact -> HTML p Action
+renderContacts contacts =
+  div
+    [ classNames [ "p-1" ] ]
+    [ h2
+        [ classNames h2Classes ]
+        [ text "Contacts" ]
+    , hr_
+    , case head contacts of
+        Nothing -> p_ [ text "You do not have any contacts." ]
+        _ ->
+          ul
+            [ classNames [ "mt-1" ] ]
+            $ contactLi
+            <$> sort contacts
+    , div
+        [ classNames [ "absolute", "bottom-1", "left-1", "right-1" ] ]
+        [ button
+            [ classNames $ buttonClasses <> [ "w-full", "px-1", "flex", "justify-between", "items-center", "bg-green", "text-white" ]
+            , onClick $ const $ Just $ ToggleNewContactCard
+            ]
+            [ span
+                [ classNames [ "mr-0.5" ] ]
+                [ text "Create new contact" ]
+            , Icon.personAdd
+            ]
+        ]
+    ]
+  where
+  contactLi contact =
+    li
+      [ classNames [ "mt-1", "hover:cursor-pointer", "hover:text-green" ]
+      , onClick $ const $ Just $ ToggleEditContactCard contact
+      ]
+      [ text $ view _nickname contact ]
+
+renderNewContact :: forall p. Array Contact -> Contact -> HTML p Action
+renderNewContact contacts newContact =
+  let
+    key = view _key newContact
+
+    nickname = view _nickname newContact
+
+    mKeyError = keyError key contacts
+
+    mNicknameError = nicknameError nickname contacts
+  in
+    div
+      [ classNames [ "flex", "flex-col" ] ]
+      [ h2
+          [ classNames h2Classes ]
+          [ text "Create new contact" ]
+      , input
+          [ type_ InputText
+          , classNames $ textInputClasses <> toggleWhen (mNicknameError == Nothing) "border-green" "border-red"
+          , placeholder "Nickname"
+          , value nickname
+          , onValueInput $ Just <<< SetNewContactNickname
```

Comment:
if you use `<<<` once you might as well use it twice `onValueInput <<< Just <<< SetNewContactNickname`

Reply:
That doesn't work. `onValueInput` takes a function that returns a `Maybe`.

Reply:
ah of course